// grunt-html2pdf/tasks/html2pdf.js
/*
 * grunt-html2pdf
 * https://github.com//grunt-html2pdf
 *
 * Copyright (c) 2014 Grégory Pelletey
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {
  var phantom = require('node-phantom-simple');
  var st = require('st');
  var http = require('http');
  var async = require('async');

  process.setMaxListeners(0);
  grunt.registerMultiTask('html2pdf', 'Create a PDF from your HTML page, useful for book or invoice generation.', function() {
    var done = this.async();
    var options = this.options({
      path: __dirname + '/pdf',
      remote: {
        files: [
          {src: "http://www.google.com", dest: "google.pdf"}
        ]
      },
      local: {
        path: "./dist",
        port: 8080,
        files: [
          {src: "index.html", dest: "page.pdf"}
        ]
      },
      format: ['A4']
    });

    // Core screenshot function using phamtonJS
    var screenshot = function(opts, cb) {
      var format = opts.format;
      var type = opts.type;
      var path = opts.path;
      var src = opts.src;
      var dest = opts.dest;
      var size;
      var ext;

      phantom.create(function(err, ph) {
        if (err) {
          grunt.fail.warn(err.message);
          return;
        }
        return ph.createPage(function(err, page) {
          ext = dest.substr(-4);
          grunt.log.writeln('Extension ' + ext);
          if (ext === ".pdf") {
            page.set('paperSize', {
              format:format,
              orientation:"portrait", 
              margin:"1cm"
            });
          }
          page.set('zoomFactor', 1);
          return page.open(src, function(err, status) {
            var target = type + '-' + format + '-' + dest;

            // Background problem under self-host server
            page.evaluate(function() {
              var style = document.createElement('style');
              var text = document.createTextNode('body { background: #fff }');
              style.setAttribute('type', 'text/css');
              style.appendChild(text);
              document.head.insertBefore(style, document.head.firstChild);
            });
            page.render(path + '/' + target, function() {
              grunt.log.writeln('Save as PDF ' + target);
              ph.exit();
              cb();
            });            
          });
        });
      }, {
        phantomPath: require('phantomjs').path
      });
    };

    // At least local or remote url should be assigned
    if (!options.remote && !options.local) {
      grunt.fail.fatal('At least need one either remote or local url');
    }

    var hasRemote = false;
    if (options.remote) {
      hasRemote = true;
      async.eachSeries(options.remote.files, function(file, outerCb) {
        async.eachSeries(options.format, function(format, cb) {
          screenshot({
            path: options.path,
            type: "remote",
            format: format,
            src: file.src,
            dest: file.dest,
          }, function() {
            cb();
          });
        }, function() {
          outerCb();
        });
      }, function() {
        grunt.event.emit('finish', 'remote');
      });
    }

    var hasLocal = false;
    if (options.local) {
      hasLocal = true;
      async.eachSeries(options.local.files, function(file, outerCb) {
        var mount = st({path: options.local.path, index: file.src});
        var server = http.createServer(function(req, res) {
          mount(req, res);
        }).listen(options.local.port, function() {
          async.eachSeries(options.viewport, function(view, cb) {
            screenshot({
              path: options.path,
              type: 'local',
              viewport: view, 
              src: 'http://localhost:' + options.local.port + '/' + file.src,
              dest: file.dest
            }, function() {
              cb();
            });
          }, function() {
            server.close();
            outerCb();
          });
        });
      }, function() {
        grunt.event.emit('finish', 'local');
      });
    }

    // Listen event to decide when can stop the task 
    grunt.event.on('finish', function(eventType) {
      if (eventType === 'remote') {
        hasRemote = false;
      }
      if (eventType === 'local') {
        hasLocal = false;
      }
      if (!hasRemote && !hasLocal) {
        done();
      }
    });
  });
};
